### ISSUE TITLE
Docs: findOrCreate broken with high frequency calls

### ISSUE THREAD

**Issue by anasyb** [balderdashy/waterline#929]
`findOrCreate` is resulting in multiple created records when multiple requests with the same criteria come at the same time.

This only happens when no record matching the criteria exists already... and say 7 requests come in together... they will result in 7 created records. Subsequent requests, no matter how frequent, do not result in more records, obviously since there is already a match that had been persisted to the db. This causes a mess: sometimes you get one record and sometimes you get multiple ones... factor in associations and you can see how messy it gets :(

Seems like the queue isn't checking for uniques.

With this and https://github.com/balderdashy/waterline/issues/877, I think it should be a standard in waterline's tests to do couple of rounds going in the same tick. These two issues were really very tough for me to debug since they only happen sometimes.

**Comment by devinivy** [comment]
Unless the adapter has its own `findOrCreate` method, the most waterline can do is look for the record, then create it if it doesn't currently exist in the DB.  Do you think `findOrCreate` should have its own cache of pending queries?  There's still a scalability issue when waterline is run on multiple different machines, say.  The most I can imagine doing is using some sort of lock on the database-level to guarantee writes have finished.  This doesn't currently seem to be a same-tick timing issue like update nested associations had.  As far as I can tell, it's simply a latency issue.  I expect it would be best solved above or below waterline, at the app level or at the database level respectively.

**Comment by devinivy** [comment]
Or there could be a giant feature requests here for waterline to create its own locks based upon db ops shared in some caching layer.  That is, redis/memcached could store ops that waterline has made on various tables, then cause blocking in the application to make sure everything is resolved.  This would be related to transaction support, requested at #755 .

**Comment by anasyb** [comment]
I see.. yes perhaps it doesn't belong in here unless a caching layer is there. I didn't check the code, but for some reason I assumed the way `findOrCreate` is implemented by waterline involves ops queueing, in which case what I would do is simple check for `findOrCreate` duplicates based on the criteria and trigger blocking in that case. 

I do think there is a bit of room for improving the docs there to emphasize that waterline isn't doing anything fancy for `findOrCreate`. Just checked mongodb docs for `findAndModify` (which can behave like `findOrCreate` when the `upsert` flag is on).. and it really did a good job at highlighting this latency issue: http://docs.mongodb.org/manual/reference/method/db.collection.findAndModify/#upsert-and-unique-index

I think the guys at mongo won't mind if sails copy/paste that section :P

**Comment by devinivy** [comment]
Good find!  We'll plan to add that to the docs.

**Comment by Globegitter** [comment]
@anasyb another temporary solution might be to do the calls in sync. Depending on performance and the rest of your code this may or may not be feasible.

But you could either use [babeljs](http://babeljs.io/) to add `async/await` support, or probably better and simpler yet just chain the `findOrCreate` returned promises as seen here: https://github.com/balderdashy/waterline/#query-methods

But yes, this should definitely be documented.

**Comment by anasyb** [comment]
Thanks @Globegitter! Those duplicate findOrCreate originate from independent API requests. If I understood you correctly that solution doesn't apply for my case.

I have very high find to create ratio, maybe 99% of the time I will be finding, and only 1% creating. For now I restructured the db and used the unique indices trick. When this issue happens, the db will return a duplicate key error. 

Since this seems to be the goto strategy for atomic findOrCreate, what I would love for waterline to have is a smarter findOrCreate which detects such latency case by watching for a duplicate keys error and i.e. automatically re-attempt. This way the developer don't have to worry about all the boilerplate code to check for duplicate key error, checking if it was caused by the insert criteria, and re-attempting the findOrCreate. Maybe we should open a separate feature request issue for that?
